### ISSUE TITLE
[V6] Field defaultValue not initialized in Store, nor passed to onSubmit

### ISSUE THREAD

**Issue by ebwill** [erikras/redux-form#1341]
I would think that at the very least field defaults should be passed to the `onSubmit` handler to reflect the actual values of the fields. If not, it places an unnecessary burden on the `onSubmit` handler to be aware of the default values, in the case that they are needed (not every API allows sparse updates).

**Comment by ebwill** [comment]
To clarify, I am seeing this behavior on v6 rc.2. Not sure if it exists in v5.

**Comment by olslash** [comment]
+1 @erikras I'm trying to figure out a good way around this issue -- right now the only thing defaultValue does is overwrite `props.input.value` in the case that no value is given (defaulting to empty string)

I'd like to see default values be more integrated into the data flow. They should be in the store just like if they were set by onChange, that way we don't have to explicitly check and re-populate the store state with them in a submit handler.

**Comment by staskorz** [comment]
+1. Not only it's important for onSubmit(), but also for validate().

**Comment by alyrik** [comment]
+1. It's become initialized in Store only after touching and blurring input.

**Comment by ktaras** [comment]
+1. Have the same on v6-rc.5. Any thoughts how to make a hotfix?

**Comment by erikras** [comment]
Thanks guys. Will investigate. 🔍

**Comment by msaglietto** [comment]
Same is happening to me on rc.4

This is not really is going to apply to all (because I did it just for one field) but what I did as a workaround is on componentWillReceiveProps check for the initialValues of that field and if is not set dispatch a change with the default value .. 
```
componentWillReceiveProps(nextProps) {
  if (!nextProps.initialValues || !nextProps.initialValues.state) {
      this.props.dispatch(this.props.change('state', 'inactive'));
    }
}
```

If you are on rc.5 it seams initialValues are not longer on the props so you have to map the field to a prop or something like that

**Comment by erikras** [comment]
> If you are on rc.5 it seams initialValues are not longer on the props so you have to map the field to a prop or something like that

@msaglietto, are you referring to #1570?

**Comment by msaglietto** [comment]
@erikras yes but no,  because I am not using mapStateToProps to get initialValues on rc.4 ... I did not tried to get initialValues using mapStateToProps on rc.5 ..

What I mean on rc.5 it may be something like mapStateToProps of 'state' .. to know if it was set .. and replace nextProps.initialValues.state by nextProps.state

**Comment by erikras** [comment]
I think what we have here is a failure to communicate. The `defaultValue` prop is only used to answer the question: "__What value should be passed to the field rendering component if the value in the Redux store is `undefined`?__" It is useful when your component expects a specific type, like a `Number` or a `Date`. It has nothing to do with actually setting the value in the form or submission or validation.

If you want to populate a value in the form, you will need to use `initialValues`.

Does that make sense? Perhaps `defaultValue` is the wrong name or the docs need to be more clear.

**Comment by brotzky** [comment]
Something like this doesn't stop working for me in `rc5`

```JavaScript
myForm = reduxForm({
  form: 'myForm',
  destroyOnUnmount: false,
  initialValues,
  validate,
})(myForm);
```

`initialValues` wouldn't populate the form anymore. Reverted back to `rc4` and had no issues.

**Comment by erikras** [comment]
Thanks, @brotzky. I wonder how that got broken.... 🔍 Investigating...

**Comment by erikras** [comment]
> I'd like to see default values be more integrated into the data flow. They should be in the store just like if they were set by onChange, that way we don't have to explicitly check and re-populate the store state with them in a submit handler.

@olslash Perhaps a separate `initialValue` prop on each field? That's a little disgusting because it would interfere with all the existing `initialValues` logic, and each field would have to dispatch an action to tell Redux about its initial value. 😕

**Comment by ebwill** [comment]
@erikras After fiddling around with some solutions myself the flow is admittedly disgusting. However, the current behavior just doesn't make sense, where defaultValue is supported but gets updated in the store only on blur. Best solution might be to drop defaultValue altogether and encourage people to use `initialValues` instead

**Comment by erikras** [comment]
@ebwill I think you might have an excellent point.

**Comment by olslash** [comment]
@erikras I think I agree that it's a little overblown, now.

In my case I'm generating a form with a bunch of fields which aren't known or assembled until runtime, so it felt easier to define the default value of a particular type of Field component on that field directly, based on runtime info, rather than supplying initialValues on the reduxForm decorator and worrying about matching the shape of the initialValues object to the final shape of the generated form.

That said, I've since started to do the work to make everything happen with initialValues, and it's really not that bad.

**Comment by olslash** [comment]
The only thing is that your defaultValue is indeed going to be set into state once you blur that field for the first time, so it feels like a feature that's *almost* but not quite working in the way I originally imagined

**Comment by erikras** [comment]
Okay, `defaultValue` has been laid to rest by #1593. It was too confusing. I will comment here again when the change has been released.

**Comment by naw** [comment]
I thought `defaultValue` was useful for the scenarios you [outlined here](https://github.com/erikras/redux-form/issues/1341#issuecomment-241055909), but I definitely agree it could be a source of confusion, perhaps because the world "default" carries a lot of connotations.

For custom form controls that use non-string values, I suppose it's practical to have a wrapper component that intercepts the empty string value provided by `redux-form` and substitutes the correct "default" value for that particular type of form control:

```javascript
DateInputWrapper = (props) => <DateInput {...props.input} value={props.input.value || myDefaultDateObject} />
```

I'd suggest that `initialValues` is a poor place for these kinds of "defaults" because each individual form should not need knowledge about what the "default/empty" value is for a variety of custom form controls.

**Comment by erikras** [comment]
Published in [`v6.0.1`](https://github.com/erikras/redux-form/releases/tag/v6.0.1).

**Comment by edyn-care** [comment]
Hey @erikras, without defaultValue working, what do I use to initialise new fields (not covered in initialValues)? Thanks
